lib/OpenQA/Worker/Settings.pm

```diff
@@ -0,0 +1,106 @@
+# Copyright (C) 2019 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+package OpenQA::Worker::Settings;
+use Mojo::Base -base;
+use Mojo::Util 'trim';
+
+use Config::IniFiles;
+
+has 'global_settings';
+has 'webui_hosts';
+has 'webui_host_specific_settings';
+
+sub new {
+    my ($class, $instance_number, $cli_options) = @_;
+    $cli_options //= {};
+
+    my $cfg = Config::IniFiles->new(-file => ($ENV{OPENQA_CONFIG} || '/etc/openqa') . '/workers.ini');
+
+    # read global settings from config
+    my %global_settings;
+    for my $section ('global', $instance_number) {
+        if ($cfg && $cfg->SectionExists($section)) {
+            for my $set ($cfg->Parameters($section)) {
+                $global_settings{uc $set} = trim $cfg->val($section, $set);
+            }
+        }
+    }
+
+    # read global settings from environment variables
+    if (my $log_dir = $ENV{OPENQA_WORKER_LOGDIR}) {
+        $global_settings{LOG_DIR} = $log_dir;
+    }
```

Comment:
what's the problem with just unconditionally setting the value?:

```perl
$global_settings{LOG_DIR} = $ENV{OPENQA_WORKER_LOGDIR};
```

or would you have a problem with auto-vivification on the second level later on?

Reply:
I don't want to override `$global_settings{LOG_DIR}` in case `$ENV{OPENQA_WORKER_LOGDIR}` is empty. Without the `if` one wouldn't be able to set the `LOG_DIR` in `workers.ini` because it would always be overridden with the value from the environment variable.